File: ingestion/transform/dimensions.py
"""Load reference and dimension tables; return IngestState with jolpica_id → internal_id maps."""

import pathlib
import sqlite3

from ingestion.transform import IngestState
from ingestion.transform.slugs import deduplicate, driver_slug, slugify
from ingestion.transform.util import as_bool, as_float, as_int, read_csv


def load_reference(con: sqlite3.Connection, dump_dir: pathlib.Path) -> None:
    _load_championship_systems(con, dump_dir)
    _load_point_systems(con, dump_dir)
    _load_base_teams(con, dump_dir)


def load_dimensions(con: sqlite3.Connection, dump_dir: pathlib.Path) -> IngestState:
    return IngestState(
        seasons  = _load_seasons(con, dump_dir),
        circuits = _load_circuits(con, dump_dir),
        drivers  = _load_drivers(con, dump_dir),
        teams    = _load_teams(con, dump_dir),
    )


# ---------------------------------------------------------------------------

def _load_championship_systems(con: sqlite3.Connection, dump_dir: pathlib.Path) -> None:
    rows = read_csv(dump_dir, "formula_one_championshipsystem.csv")
    con.executemany("""
        INSERT INTO championship_systems
          (id, jolpica_api_id, name, reference,
           driver_best_results, driver_season_split, eligibility,
           team_best_results, team_points_per_session, team_season_split)
        VALUES (?,?,?,?,?,?,?,?,?,?)
    """, [(
        int(r["id"]), r["api_id"], r["name"], r["reference"],
        int(r["driver_best_results"]), int(r["driver_season_split"]),
        int(r["eligibility"]),
        int(r["team_best_results"]), int(r["team_points_per_session"]),
        int(r["team_season_split"]),
    ) for r in rows])


def _load_point_systems(con: sqlite3.Connection, dump_dir: pathlib.Path) -> None:
    rows = read_csv(dump_dir, "formula_one_pointsystem.csv")
    con.executemany("""
        INSERT INTO point_systems
          (id, jolpica_api_id, name, reference, partial,
           driver_position_points, driver_fastest_lap,
           team_position_points, team_fastest_lap,
           is_double_points, shared_drive)
        VALUES (?,?,?,?,?,?,?,?,?,?,?)
    """, [(
        int(r["id"]), r["api_id"], r["name"], r.get("reference", ""),
        as_int(r, "partial") or 0,
        r.get("driver_position_points", ""),
        as_float(r, "driver_fastest_lap") or 0.0,
        r.get("team_position_points", ""),
        as_float(r, "team_fastest_lap") or 0.0,
        as_bool(r, "is_double_points"),
        as_int(r, "shared_drive") or 0,
    ) for r in rows])


def _load_base_teams(con: sqlite3.Connection, dump_dir: pathlib.Path) -> None:
    rows = read_csv(dump_dir, "formula_one_baseteam.csv")
    if not rows:
        return  # CSV is empty in the current dump
    con.executemany(
        "INSERT INTO base_teams (id, jolpica_api_id, name) VALUES (?,?,?)",
        [(int(r["id"]), r["api_id"], r.get("name", "")) for r in rows],
    )


def _load_seasons(con: sqlite3.Connection, dump_dir: pathlib.Path) -> dict:
    """Returns {jolpica_id: year}."""
    rows = read_csv(dump_dir, "formula_one_season.csv")
    con.executemany("""
        INSERT INTO seasons (year, jolpica_id, jolpica_api_id, championship_system_id, wikipedia)
        VALUES (?,?,?,?,?)
    """, [(
        int(r["year"]), int(r["id"]), r["api_id"],
        as_int(r, "championship_system_id"),
        r.get("wikipedia", ""),
    ) for r in rows])
    return {int(r["id"]): int(r["year"]) for r in rows}


def _load_circuits(con: sqlite3.Connection, dump_dir: pathlib.Path) -> dict:
    """Returns {jolpica_id: internal_id}."""
    rows = read_csv(dump_dir, "formula_one_circuit.csv")
    existing_slugs: set[str] = set()
    records = []
    for r in rows:
        base = r.get("reference") or slugify(r["name"])
        slug = deduplicate(base, existing_slugs)
        existing_slugs.add(slug)
        records.append((
            slug, int(r["id"]), r["api_id"],
            r["name"], r.get("locality", ""), r.get("country", ""),
            r.get("country_code", ""),
            as_float(r, "latitude"),
            as_float(r, "longitude"),
            as_int(r, "altitude"),
            r.get("reference", ""), r.get("wikipedia", ""),
        ))
    con.executemany("""
        INSERT INTO circuits
          (slug, jolpica_id, jolpica_api_id, name, locality, country, country_code,
           latitude, longitude, altitude, reference, wikipedia)
        VALUES (?,?,?,?,?,?,?,?,?,?,?,?)
    """, records)
    return {
        row["jolpica_id"]: row["id"]
        for row in con.execute("SELECT id, jolpica_id FROM circuits")
    }


def _load_drivers(con: sqlite3.Connection, dump_dir: pathlib.Path) -> dict:
    """Returns {jolpica_id: internal_id}."""
    rows = sorted(
        read_csv(dump_dir, "formula_one_driver.csv"),
        key=lambda r: (r["surname"], r["forename"], r["id"]),
    )
    existing_slugs: set[str] = set()
    records = []
    for r in rows:
        base = driver_slug(r["forename"], r["surname"])
        slug = deduplicate(base, existing_slugs)
        existing_slugs.add(slug)
        records.append((
            slug, int(r["id"]), r["api_id"],
            r["forename"], r["surname"], f"{r['forename']} {r['surname']}",
            r.get("abbreviation") or None,
            as_int(r, "permanent_car_number"),
            r.get("country_code") or None,
            r.get("nationality") or None,
            r.get("date_of_birth") or None,
            r.get("reference") or None,
            r.get("wikipedia") or None,
        ))
    con.executemany("""
        INSERT INTO drivers
          (slug, jolpica_id, jolpica_api_id, forename, surname, full_name,
           abbreviation, permanent_car_number, country_code, nationality,
           date_of_birth, reference, wikipedia)
        VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)
    """, records)
    return {
        row["jolpica_id"]: row["id"]
        for row in con.execute("SELECT id, jolpica_id FROM drivers")
    }


def _load_teams(con: sqlite3.Connection, dump_dir: pathlib.Path) -> dict:
    """Returns {jolpica_id: internal_id}."""
    rows = read_csv(dump_dir, "formula_one_team.csv")
    existing_slugs: set[str] = set()
    records = []
    for r in rows:
        base = r.get("reference") or slugify(r["name"])
        slug = deduplicate(base, existing_slugs)
        existing_slugs.add(slug)
        records.append((
            slug, int(r["id"]), r["api_id"],
            as_int(r, "base_team_id"),
            r["name"],
            r.get("country_code") or None,
            r.get("nationality") or None,
            r.get("primary_color") or None,
            r.get("reference") or None,
            r.get("wikipedia") or None,
        ))
    con.executemany("""
        INSERT INTO teams
          (slug, jolpica_id, jolpica_api_id, base_team_id, name,
           country_code, nationality, primary_color, reference, wikipedia)
        VALUES (?,?,?,?,?,?,?,?,?,?)
    """, records)
    return {
        row["jolpica_id"]: row["id"]
        for row in con.execute("SELECT id, jolpica_id FROM teams")
    }

File: ingest/transform/dimensions.ts
import type { LibSQLDatabase } from 'drizzle-orm/libsql';
import { readCsv, asInt, asFloat, asBool } from '../csv';
import { slugify, driverSlug, deduplicate } from '../slugs';
import { emptyState, type IngestState } from '../state';
import * as schema from '../../src/db/schema';

const CHUNK = 500;

async function insertChunked<T>(
  db: LibSQLDatabase<typeof schema>,
  table: Parameters<typeof db.insert>[0],
  rows: T[],
): Promise<void> {
  for (let i = 0; i < rows.length; i += CHUNK) {
    await db.insert(table).values(rows.slice(i, i + CHUNK) as [T, ...T[]]);
  }
}

export async function loadReference(
  db: LibSQLDatabase<typeof schema>,
  dumpDir: string,
): Promise<void> {
  await loadChampionshipSystems(db, dumpDir);
  await loadPointSystems(db, dumpDir);
  await loadBaseTeams(db, dumpDir);
}

export async function loadDimensions(
  db: LibSQLDatabase<typeof schema>,
  dumpDir: string,
): Promise<IngestState> {
  const state = emptyState();
  state.seasons = await loadSeasons(db, dumpDir);
  state.circuits = await loadCircuits(db, dumpDir);
  state.drivers = await loadDrivers(db, dumpDir);
  state.teams = await loadTeams(db, dumpDir);
  return state;
}

async function loadChampionshipSystems(db: LibSQLDatabase<typeof schema>, dumpDir: string) {
  const rows = readCsv(dumpDir, 'formula_one_championshipsystem.csv');
  const records = rows.map(r => ({
    id: parseInt(r.id, 10),
    jolpicaApiId: r.api_id,
    name: r.name,
    reference: r.reference || null,
    driverBestResults: asInt(r, 'driver_best_results') ?? 0,
    driverSeasonSplit: asInt(r, 'driver_season_split') ?? 0,
    eligibility: asInt(r, 'eligibility') ?? 1,
    teamBestResults: asInt(r, 'team_best_results') ?? 0,
    teamPointsPerSession: asInt(r, 'team_points_per_session') ?? 0,
    teamSeasonSplit: asInt(r, 'team_season_split') ?? 0,
  }));
  await insertChunked(db, schema.championshipSystems, records);
}

async function loadPointSystems(db: LibSQLDatabase<typeof schema>, dumpDir: string) {
  const rows = readCsv(dumpDir, 'formula_one_pointsystem.csv');
  const records = rows.map(r => ({
    id: parseInt(r.id, 10),
    jolpicaApiId: r.api_id,
    name: r.name,
    reference: r.reference || null,
    partial: asInt(r, 'partial') ?? 0,
    driverPositionPoints: r.driver_position_points || null,
    driverFastestLap: asFloat(r, 'driver_fastest_lap') ?? 0,
    teamPositionPoints: r.team_position_points || null,
    teamFastestLap: asFloat(r, 'team_fastest_lap') ?? 0,
    isDoublePoints: asBool(r, 'is_double_points'),
    sharedDrive: asInt(r, 'shared_drive') ?? 0,
  }));
  await insertChunked(db, schema.pointSystems, records);
}

async function loadBaseTeams(db: LibSQLDatabase<typeof schema>, dumpDir: string) {
  const rows = readCsv(dumpDir, 'formula_one_baseteam.csv');
  if (!rows.length) return;
  const records = rows.map(r => ({
    id: parseInt(r.id, 10),
    jolpicaApiId: r.api_id,
    name: r.name || null,
  }));
  await insertChunked(db, schema.baseTeams, records);
}

async function loadSeasons(db: LibSQLDatabase<typeof schema>, dumpDir: string): Promise<Map<number, number>> {
  const rows = readCsv(dumpDir, 'formula_one_season.csv');
  const records = rows.map(r => ({
    year: parseInt(r.year, 10),
    jolpicaId: parseInt(r.id, 10),
    jolpicaApiId: r.api_id,
    championshipSystemId: asInt(r, 'championship_system_id'),
    wikipedia: r.wikipedia || null,
  }));
  await insertChunked(db, schema.seasons, records);
  return new Map(rows.map(r => [parseInt(r.id, 10), parseInt(r.year, 10)]));
}

async function loadCircuits(db: LibSQLDatabase<typeof schema>, dumpDir: string): Promise<Map<number, number>> {
  const rows = readCsv(dumpDir, 'formula_one_circuit.csv');
  const existingSlugs = new Set<string>();
  const records = rows.map(r => {
    const base = r.reference || slugify(r.name);
    const slug = deduplicate(base, existingSlugs);
    existingSlugs.add(slug);
    return {
      slug,
      jolpicaId: parseInt(r.id, 10),
      jolpicaApiId: r.api_id,
      name: r.name,
      locality: r.locality || null,
      country: r.country || null,
      countryCode: r.country_code || null,
      latitude: asFloat(r, 'latitude'),
      longitude: asFloat(r, 'longitude'),
      altitude: asInt(r, 'altitude'),
      reference: r.reference || null,
      wikipedia: r.wikipedia || null,
    };
  });
  await insertChunked(db, schema.circuits, records);
  const inserted = await db.select({ id: schema.circuits.id, jolpicaId: schema.circuits.jolpicaId }).from(schema.circuits);
  return new Map(inserted.map(r => [r.jolpicaId, r.id]));
}

async function loadDrivers(db: LibSQLDatabase<typeof schema>, dumpDir: string): Promise<Map<number, number>> {
  const rows = readCsv(dumpDir, 'formula_one_driver.csv');
  // Sort for stable slug collision suffixes across re-runs
  rows.sort((a, b) => {
    const s = a.surname.localeCompare(b.surname);
    if (s !== 0) return s;
    const f = a.forename.localeCompare(b.forename);
    if (f !== 0) return f;
    return parseInt(a.id, 10) - parseInt(b.id, 10);
  });
  const existingSlugs = new Set<string>();
  const records = rows.map(r => {
    const base = driverSlug(r.forename, r.surname);
    const slug = deduplicate(base, existingSlugs);
    existingSlugs.add(slug);
    return {
      slug,
      jolpicaId: parseInt(r.id, 10),
      jolpicaApiId: r.api_id,
      forename: r.forename,
      surname: r.surname,
      fullName: `${r.forename} ${r.surname}`,
      abbreviation: r.abbreviation || null,
      permanentCarNumber: asInt(r, 'permanent_car_number'),
      countryCode: r.country_code || null,
      nationality: r.nationality || null,
      dateOfBirth: r.date_of_birth || null,
      reference: r.reference || null,
      wikipedia: r.wikipedia || null,
    };
  });
  await insertChunked(db, schema.drivers, records);
  const inserted = await db.select({ id: schema.drivers.id, jolpicaId: schema.drivers.jolpicaId }).from(schema.drivers);
  return new Map(inserted.map(r => [r.jolpicaId, r.id]));
}

async function loadTeams(db: LibSQLDatabase<typeof schema>, dumpDir: string): Promise<Map<number, number>> {
  const rows = readCsv(dumpDir, 'formula_one_team.csv');
  const existingSlugs = new Set<string>();
  const records = rows.map(r => {
    const base = r.reference || slugify(r.name);
    const slug = deduplicate(base, existingSlugs);
    existingSlugs.add(slug);
    return {
      slug,
      jolpicaId: parseInt(r.id, 10),
      jolpicaApiId: r.api_id,
      baseTeamId: asInt(r, 'base_team_id'),
      name: r.name,
      countryCode: r.country_code || null,
      nationality: r.nationality || null,
      primaryColor: r.primary_color || null,
      reference: r.reference || null,
      wikipedia: r.wikipedia || null,
    };
  });
  await insertChunked(db, schema.teams, records);
  const inserted = await db.select({ id: schema.teams.id, jolpicaId: schema.teams.jolpicaId }).from(schema.teams);
  return new Map(inserted.map(r => [r.jolpicaId, r.id]));
}
